fw/tls.c

```diff
@@ -664,10 +665,10 @@ tfw_tls_conn_init(TfwConn *c)
 		goto err_cleanup;
 	}
 
-	if (TFW_FSM_TYPE(c->proto.type) == TFW_FSM_H2)
+	if (TFW_FSM_TYPE(c->proto.type) == TFW_FSM_H2 ||
+	    TFW_FSM_TYPE(c->proto.type) == TFW_FSM_H2_HTTPS)
```

Comment:
Another overhead for HTTP1, we initialize h2 context(including hpack) even for http1. Would be better to do it, when we know protocol.

@krizhanovsky Why do we init higher layer protocol context(http and http2) before finishing lower layer connection establishing, I mean TLS. Is there a reason or just implemented this way?

Reply:
Yes, this is kind of reversion of https://github.com/tempesta-tech/tempesta/commit/3c97cb23d - we should initialize HTTP/2 context only for HTTP/2 connections. Note also https://github.com/tempesta-tech/tempesta/issues/1422 - another HTTP/2 -related regression in HTTP/1.

@const-t Frankly, I don't remember why do we initialize HTTP/2 context before finishing TLS handshake. It makes sense to allocate memory necessary for HTTP/2 early to improve spacial locality and reduce number of allocator calls. But for now for me it seems that we can initialize the context later... Need to check commits history - I suppose this point should have been discussed at some point.

Reply:
> Another overhead for HTTP1, we initialize h2 context(including hpack) even for http1. Would be better to do it, when we know protocol.

I replace initialization h2 context for TFW_FSM_H2_HTTPS to tfw_http_msg_process